NLP/main/extract_display_quotes.py

```diff
@@ -0,0 +1,531 @@
+            text_id: the text_id of the text you wish to display
+            show_what: options to display speakers, quotes or named entities
+            inc_ent: a list containing the named entities to be extracted from the text, 
+                     e.g., ['ORG','PERSON','GPE','NORP','FAC','LOC']
+            save_to_html: option to save the display into an html file
+        '''
+        # formatting options
+        TPL_SPAN = '''
+        <span style="font-weight: bold; display: inline-block; position: relative; 
+        line-height: 55px">
+            {text}
+            {span_slices}
+            {span_starts}
+        </span>
+        '''
+        
+        TPL_SPAN_SLICE = '''
+        <span style="background: {bg}; top: {top_offset}px; height: 4px; left: -1px; width: calc(100% + 2px); position: absolute;">
+        </span>
+        '''
+        
+        TPL_SPAN_START = '''
+        <span style="background: {bg}; top: {top_offset}px; height: 4px; border-top-left-radius: 3px; border-bottom-left-radius: 3px; left: -1px; width: calc(100% + 2px); position: absolute;">
+            <span style="background: {bg}; z-index: 10; color: #000; top: -0.5em; padding: 2px 3px; position: absolute; font-size: 0.6em; font-weight: bold; line-height: 1; border-radius: 3px">
+                {label}{kb_link}
+            </span>
+        </span>
+        '''
+        
+        colors = {'QUOTE': '#66ccff', 'SPEAKER': '#66ff99'}
+        options = {'ents': ['QUOTE', 'SPEAKER'], 
+                   'colors': colors, 
+                   'top_offset': 42,
+                   'template': {'span':TPL_SPAN,
+                               'slice':TPL_SPAN_SLICE,
+                               'start':TPL_SPAN_START},
+                   'span_label_offset': 14,
+                   'top_offset_step':14}
+        
+        # get the spaCy text 
+        doc = text_df.loc[text_id, 'spacy_text']
+        
+        # create a mapping dataframe between the character index and token index from the spacy text.
+        loc2tok_df = pd.DataFrame([(t.idx, t.i) for t in doc], columns = ['loc', 'token'])
+    
+        # get the quotes and speakers indexes
+        locs = {
+            'QUOTE': quotes_df[quotes_df['text_id']==text_id]['quote_index'].tolist(),
+            'SPEAKER': set(quotes_df[quotes_df['text_id']==text_id]['speaker_index'].tolist())
+        }
+    
+        # create the displaCy code to visualise quotes and speakers
+        my_code_list = ['doc.spans["sc"] = [', ']']
+        
+        for key in locs.keys():
+            for loc in locs[key]:
+                if loc!=(0,0):
+                    # Find out all token indices that falls within the given span (variable loc)
+                    selTokens = loc2tok_df.loc[(loc[0]<=loc2tok_df['loc']) & (loc2tok_df['loc']<loc[1]), 'token'].tolist()
+                    
+                    # option to display named entities only
+                    if show_what==['NAMED ENTITIES']:
```

Comment:
The following lines are duplicated, should be able to make them a reusable function/module.

Reply:
function created